### PR TITLE
2014 refactor variant count on hand

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Part of #2014.

This prepares the inventory report and maybe more for the Spree upgrade. It also fixes a bug in filtering stocked variants and removes some unused code.

#### What should we test?

- Check that the inventory report displays the right list of variants.
- Also check shops for correct display of item availability.
- And sanity check stock management.

#### Release notes

Fixed bug in inventory report which ignored the distributor or on-demand setting for stock calculation. This may slow down the inventory report generation.

#### How is this related to the Spree upgrade?

This refactors some code to be compatible with Spree 2.0.

### CODE REVIEW

**Comment by mkllnk** [comment]
@sauloperez I just noticed that I repeated your commit 0cbd967007d05dee77d440e5f2f406ef70eaef2d. @Matt-Yorkley added that line in a503b5e2d0b1a703f1bc5527a45241d7bb7776e4. I guess that was a merge conflict mistake.

File: lib/open_food_network/scope_variant_to_hub.rb [hunk 1]
```diff
@@ -21,6 +20,14 @@ def price_in(currency)
         Spree::Price.new(amount: price, currency: currency)
       end
 
+      # Old Spree has the same logic as here and doesn't need this override.
+      # But we need this to use VariantOverrides with Spree 2.0.
+      def in_stock?
+        return true unless Spree::Config[:track_inventory_levels]
+
+        on_demand || (count_on_hand > 0)
```
Comment (sauloperez) [inline lib/open_food_network/scope_variant_to_hub.rb]: I'm not familiar with variant overrides but I guess `count_on_hand` is a column on a database table added by OFN, right?

Comment (Matt-Yorkley) [reply lib/open_food_network/scope_variant_to_hub.rb]: It's a spree database column on both spree_product and spree_variant.

Comment (Matt-Yorkley) [reply lib/open_food_network/scope_variant_to_hub.rb]: Just a heads-up @mkllnk: `on_demand` is being removed as well in 2.0.

Comment (mkllnk) [reply lib/open_food_network/scope_variant_to_hub.rb]: @Matt-Yorkley Thanks for the heads up. I wasn't sure if I considered that. But this method is only active if the variant has a VariantOverride. The VariantOverride does have `on_demand` and `count_on_hand`. The vanilla Spree::Variant still has an `in_stock?` method in both Spree versions. So we should be safe to call it.

**Comment by mkllnk** [comment]
@sauloperez Yes, I did test that just now. But it should also be covered by our specs.

What do you mean with "too early"? This is production code, going into master.

**Comment by mkllnk** [comment]
Staged on https://staging1.openfood.com.au/.

**Comment by RachL** [comment]
@myriamboure or @sstead I'm afraid I don't know very well the inventory feature to test this PR. Can one of you have a look? I will be interested in ready the testing notes afterwards.
I've started some notes here, but let's be honest you can just use the document to save yourself the time of creating it, other than that nothing much interesting:
https://docs.google.com/document/d/1-2soDLPWbvUQNpubrDmUE6B1lrK1VlNJWNHrrpt8UM8/edit

**Comment by sstead** [comment]
@mkllnk can you clarify, are we testing all 3 reports within the Products and Inventory report?

I don't have anymore time to test today but here's where I'm up to so far...

Issues:
- In some cases, when I filter by distributor only, I'm seeing products from suppliers who have not granted that shop P-OC E2E. E.g. Sally2's Hub and Producer of Wine.
- In some cases, products are not showing for unknown reasons
- When filtering by OC, the report is filtering by the 'incoming' section of the OC, not outgoing.
- Some inventory settings aren't applying in the report - e.g. I would expect 'hidden' inventry items to be hidden on the report - not the case.
- There seems to be a bug with the Advanced settings for OCs, which makes it hard to test the impact of this setting on the OC filter in the report.

I think there's some design flaws with these reports:
- We shouldn't use the work 'inventory' unless we're specifically refering to the Inventory page. In the case of the Inventory (on hand) report, if you're filtering by a supplier only, it's got nothing to do with the Inventory page, it's referring to On Hand only. I suggest we take the word Inventory out of this report.
- It's confusing the way this report switches between looking at the BEP page (in the case of filtering by supplier only) and then switching to look at both the BEP page and the Inventory page (in the case of filtering by a distributor and a supplier). 
- In the case where I select a distributor filter, I would prefer if the other filters were then restricted to allowable options - e.g. only OCs which that distributor is in. Or only suppliers who have granted P-OC to.

https://docs.google.com/document/d/1-2soDLPWbvUQNpubrDmUE6B1lrK1VlNJWNHrrpt8UM8/edit#

**Comment by kirstenalarsen** [comment]
@sstead how sure are you that these problems are all new / due to this change? I am not at all across this, just wondering if your excellent issue detection skills have ever been applied to these reports before?

**Comment by sstead** [comment]
Yes there’s a very good chance these are prexisting, this isn’t a report I test often. Should we log the other issues for later PRs? I can test more narrowly if that’s what’s needed.

________________________________

**Comment by daniellemoorhead** [comment]
@sstead best way to tell if these are existing is to test on production, yes?

**Comment by mkllnk** [comment]
I would be very surprised if those issues are caused by my code change. I only changed the `on_hand` filter. That already fixed a bug that the `on_demand` property was ignored.

This report is much older than the new inventory feature. It probably does make sense to rename it now. Maybe stock report?

Anyway, to proceed with the Spree upgrade, I just want the report to perform as well as before. Changing the report and fixing it are other issues.

**Comment by sstead** [comment]
**Testing notes**

- Yes, the Inventory on hand report and the lettuceshare report no longer show items with <1 on hand value in Inventory.
- The shop seems to be behaving as normal
- The stock management is working as normal (pre-existing bug only)

Broader issues spotted while testing are logged in this issue - https://github.com/openfoodfoundation/openfoodnetwork/issues/2529

https://docs.google.com/document/d/1-2soDLPWbvUQNpubrDmUE6B1lrK1VlNJWNHrrpt8UM8/edit#

**Pull request merged**